luigi_pipeline/lib/model/base_mt_schema.py

```diff
@@ -0,0 +1,105 @@
+
+        @mt_annotation()
+        def a(self):
+            return 'a_val'
+
+        @mt_annotation(annotation='b', fn_require=a)
+        def b_1(self):
+            return 'b_val'
+
+    Will generate a mt with rows of {a: 'a_val', 'b': 'b_val'} if the function is called.
+    TODO: Consider changing fn_require to be a list of requirements.
+
+    :param annotation: name in the final MT. If not provided, uses the function name.
+    :param fn_require: method name strings in class that are dependencies.
+    :return:
+    """
+    def mt_prop_wrapper(func):
+        annotation_name = annotation or func.__name__
+
+        @wraps(func)
+        def wrapper(self, *args, **kwargs):
+            # Called already.
+            if wrapper.mt_prop_meta['annotated'] > 0:
+                return self
+            if fn_require:
+                getattr(self, fn_require.__name__)()
+
+            # Annotate and retiurn instance for chaining.
+            self.mt = self.mt.annotate_rows(**{annotation_name: func(self, *args, **kwargs)})
+            wrapper.mt_prop_meta['annotated'] += 1
+            return self
+
+        wrapper.mt_prop_meta = {
+            'annotated': 0,  # Counter for number of times called. Should only be 0 or 1.
+            'annotated_name': annotation_name
+        }
+        return wrapper
+    return mt_prop_wrapper
+
+
+class BaseMTSchema:
+    """
+    Main superclass that provides a Hail MT schema definition. decorate methods with @mt_annotation.
+    Allows annotations to express dependencies where dependencies are run before (and at most once).
+    NOTE: circular dependencies are not supported and not gracefully handled.
+
+    Usage example:
+        class TestSchema(BaseMTSchema):
+
+            def __init__(self):
+                super(TestSchema, self).__init__(hl.import_vcf('tests/data/1kg_30variants.vcf.bgz'))
+
+            @mt_annotation()
+            def a(self):
+                return 0
+
+            @mt_annotation(fn_require=a)
+            def b(self):
+                return self.a + 1
+
+            @mt_annotation(annotation='c', fn_require=a)
+            def c_1(self):
+                return self.a + 2
+
+    `TestSchema(mt).b().c_1().select_annotated_mt()` will annotate with {'a': 0, 'b': 1, 'c': 2}
+
+    """
+    def __init__(self, mt):
+        self.mt = mt
+
+    def all_annotation_fns(self):
+        """
+        Get all mt_annotation decorated methods using introspection.
+        :return: list of all annotation functions
+        """
+        return getmembers(self.__class__, lambda x: isfunction(x) and hasattr(x, 'mt_prop_meta'))
+
+    def annotate_all(self):
+        """
+        Iterate over all annotation functions and call them on the instance.
+        :return: instance object
+        """
+        for atn_fn in self.all_annotation_fns():
+            getattr(self, atn_fn[0])()
+        return self
+
+    def select_annotated_mt(self):
```

Comment:
Should this call `annotate_all` to ensure annotations have been applied?

Reply:
Yeah this is another case for flexibility, when we want partial annotations. So if you only annotate calls to `a` and `b`, it will only return those.

Reply:
Ah, I didn't notice the `if fn[1].mt_prop_meta['annotated'] > 0` and thought this would attempt to select all annotations even if they haven't been applied.